File: 백준/Gold/15686. 치킨 배달/치킨 배달.cc
#include <bits/stdc++.h>
using namespace std;

const int INF = 987654321;
const int MAX = 54;
int N, M, a[MAX][MAX], ret = INF;
vector<pair<int, int>> home, chicken;
vector<vector<int>> selectedChicken;

void combi(int idx, vector<int> &v) {
    if (v.size() == M) {
        selectedChicken.push_back(v);
        return;
    }
    for (int i=idx+1; i<chicken.size(); i++) {
        v.push_back(i);
        combi(i, v);
        v.pop_back();
    }
}

int main() {
    cin >> N >> M;
    for (int i=0; i<N; i++) {
        for (int j=0; j<N; j++) {
            cin >> a[i][j];
            if (a[i][j] == 1) home.push_back({i, j});
            else if (a[i][j] == 2) chicken.push_back({i, j});
        }
    }

    vector<int> v;
    combi(-1, v);

    for (vector<int> ch: selectedChicken) {
        int sum = 0;
        for (auto h: home) {
            int mn = INF;
            for (int i: ch) {
                int dist = abs(h.first - chicken[i].first) + abs(h.second - chicken[i].second);
                mn = min(mn, dist);
            }
            sum += mn;
        }
        ret = min(ret, sum);
    }
    cout << ret << '\n';
    return 0;
}

File: 백준/Gold/15686. 치킨 배달/치킨 배달.py
N, M = map(int, input().split())
graph = [list(map(int, input().split())) for _ in range(N)]

INF = 987654321
ret = INF
cList, hList = [], []
selectedChicken = []

def comb(idx: int, arr: list):
    if len(arr) == M:
        selectedChicken.append(arr[:])
        return
    
    for i in range(idx+1, len(cList)):
        arr.append(i)
        comb(i, arr)
        arr.pop()

for i in range(N):
    for j in range(N):
        if graph[i][j] == 1:
            hList.append((i, j))
        elif graph[i][j] == 2:
            cList.append((i, j))

comb(-1, [])

for ch in selectedChicken:
    total = 0
    for h in hList:
        mn = INF
        for i in ch:
            dist = abs(h[0] - cList[i][0]) + abs(h[1] - cList[i][1])
            mn = min(mn, dist)
        total += mn
    ret = min(ret, total)

print(ret)